# Mushroom

    Каждый вид идентифицируется как съедобный, ядовитый или неизвестной съедобности и не рекомендуется, кроме того присутствуют описательные характеристики:

1.	class - класс
2.	cap shape - форма шляпки
3.	cap surface - поверхность шляпки
4.	cap color - цвет шляпки
5.	bruises - повреждения
6.	odor - запах
7.	gill attachment - жаберное крепление
8.	gill spacing – плотность жабр
9.	gill size – размер жабр
10.	gill color – цвет жабр
11.	stalk shape - форма стебля
12.	stalk root - корень стебля
13.	stalk surface above ring - поверхность стебля над кольцом
14.	stalk surface below ring - поверхность стебля под кольцом
15.	stalk color above ring - цвет стебля над кольцом
16.	stalk color below ring - цвет стебля под кольцом
17.	veil type – тип покрова
18.	veil color – цвет  покрова
19.	ring number – число колец
20.	ring type – тип колец
21.	spore print color – цвет спор
22.	population - популяция
23.	habitat -  ареал


In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
%matplotlib inline

Читаем данные.

In [2]:
data = pd.read_csv('https://raw.githubusercontent.com/golyakova27/MLProject1/master/MushroomBase.csv', sep = ';', engine = 'python')

Вывод данных.

In [3]:
data

Class Cap-shape Cap-surface Cap-color Bruises Odor Gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
5        e         x           y         y       t    a               f   
6        e         b           s         w       t    a               f   
7        e         b           y         w       t    l               f   
8        p         x           y         w       t    p               f   
9        e         b           s         y       t    a               f   
10       e         x           y         y       t    l               f   
11       e         x           y         y       t    a               f   
12       e         b           s         y       t    a               f   
13       p         x           y         w       t    p               f   
14       e         x           f         n       f    n               f   
15       e         s           f         g       f    n               f   
16       e         f           f         w       f    n               f   
17       p         x           s         n       t    p               f   
18       p         x           y         w       t    p               f   
19       p         x           s         n       t    p               f   
20       e         b           s         y       t    a               f   
21       p         x           y         n       t    p               f   
22       e         b           y         y       t    l               f   
23       e         b           y         w       t    a               f   
24       e         b           s         w       t    l               f   
25       p         f           s         w       t    p               f   
26       e         x           y         y       t    a               f   
27       e         x           y         w       t    l               f   
28       e         f           f         n       f    n               f   
29       e         x           s         y       t    a               f   
...    ...       ...         ...       ...     ...  ...             ...   
8094     e         b           s         g       f    n               f   
8095     p         x           y         c       f    m               f   
8096     e         k           f         w       f    n               f   
8097     p         k           y         n       f    s               f   
8098     p         k           s         e       f    y               f   
8099     e         k           f         w       f    n               f   
8100     e         f           s         n       f    n               a   
8101     p         k           s         e       f    s               f   
8102     e         x           s         n       f    n               a   
8103     e         k           s         n       f    n               a   
8104     e         k           s         n       f    n               a   
8105     e         k           s         n       f    n               a   
8106     e         k           s         n       f    n               a   
8107     e         x           s         n       f    n               a   
8108     p         k           y         e       f    y               f   
8109     e         b           s         w       f    n               f   
8110     e         x           s         n       f    n               a   
8111     e         k           s         w       f    n               f   
8112     e         k           s         n       f    n               a   
8113     p         k           y         e       f    y               f   
8114     p         f           y         c       f    m               a   
8115     e         x          

Каждая строка представляет собой гриб, каждый столбец - его характеристика.

In [4]:
print(data)

     Class Cap-shape Cap-surface Cap-color Bruises Odor Gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
5        e         x           y         y       t    a               f   
6        e         b           s         w       t    a               f   
7        e         b           y         w       t    l               f   
8        p         x           y         w       t    p               f   
9        e         b           s         y       t    a               f   
10       e         x           y         y       t    l               f   
11       e         x           y         y       t    a               f   
12       e         b     

Вывод первых пяти строк.

In [5]:
data.head()

Class Cap-shape Cap-surface Cap-color Bruises Odor Gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  Gill-spacing Gill-size Gill-color   ...   Stalk-surface-below-ring  \
0            c         n          k   ...                          s   
1            c         b          k   ...                          s   
2            c         b          n   ...                          s   
3            c         n          n   ...                          s   
4            w         b          k   ...                          s   

  Stalk-color-above-ring Stalk-color-below-ring Veil-type Veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  Ring-number Ring-type Spore-print-color Population Habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

В полученой таблице, например:

Гриб №3 имеет характеристики:

1.	Class – класс: p(poisonous)
2.	cap shape – форма шляпки: x(convex-)
3.	cap surface - поверхность шляпки: y(scaly)
4.	cap color - цвет шляпки: w(white)
5.	bruises – повреждения: t(true)
6.	odor – запах: p(pungent)
7.	gill attachment - жаберное крепление: f(free)
8.	gill spacing – плотность жабр: c(close)
9.	gill size – размер жабр: n(narrow)
10.	gill color – цвет жабр: n(red)
11.	stalk shape - форма стебля: e(enlarging)
12.	stalk root - корень стебля: e(equal)
13.	stalk surface above ring - поверхность стебля над кольцом: s(smooth)
14.	stalk surface below ring - поверхность стебля под кольцом: s(smooth)
15.	stalk color above ring - цвет стебля над кольцом: w(white)
16.	stalk color below ring - цвет стебля под кольцом: w(white)
17.	veil type – тип покрова: p(partial)
18.	veil color – цвет  покрова: w(white)
19.	ring number – число колец: o(one)
20.	ring type – тип колец: p(pendant)
21.	spore print color – цвет спор: k(black)
22.	population – популяция: s(scattered)
23.	habitat -  ареал: u(urban)


Вывод последних пяти строк.

In [6]:
data.tail()

Class Cap-shape Cap-surface Cap-color Bruises Odor Gill-attachment  \
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     Gill-spacing Gill-size Gill-color   ...   Stalk-surface-below-ring  \
8119            c         b          y   ...                          s   
8120            c         b          y   ...                          s   
8121            c         b          n   ...                          s   
8122            c         n          b   ...                          k   
8123            c         b          y   ...                          s   

     Stalk-color-above-ring Stalk-color-below-ring Veil-type Veil-color  \
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     Ring-number Ring-type Spore-print-color Population Habitat  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[5 rows x 23 columns]

Вывод размера данных.

In [7]:
print(data.shape)

(8124, 23)


В данной таблице 23 столбца и 8124 строки (23 характеристики, 8124 гриба).

Можно посмотреть имеющиеся названия столбцов с помощью

In [8]:
print(data.columns)

Index(['Class', 'Cap-shape', 'Cap-surface', 'Cap-color', 'Bruises', 'Odor',
       'Gill-attachment', 'Gill-spacing', 'Gill-size', 'Gill-color',
       'Stalk-shape', 'Stalk-root', 'Stalk-surface-above-ring',
       'Stalk-surface-below-ring', 'Stalk-color-above-ring',
       'Stalk-color-below-ring', 'Veil-type', 'Veil-color', 'Ring-number',
       'Ring-type', 'Spore-print-color', 'Population', 'Habitat'],
      dtype='object')


Или же посмотреть значение выбранного столбца в конкретной строке таблице.

In [9]:
data.at[8122, 'Odor']

'y'

Вывод выбранного столбца.

In [10]:
data['Gill-size']

0       n
1       b
2       b
3       n
4       b
5       b
6       b
7       b
8       n
9       b
10      b
11      b
12      b
13      n
14      b
15      n
16      b
17      n
18      n
19      n
20      b
21      n
22      b
23      b
24      b
25      n
26      b
27      b
28      n
29      n
       ..
8094    b
8095    b
8096    b
8097    n
8098    n
8099    b
8100    b
8101    n
8102    b
8103    b
8104    b
8105    b
8106    b
8107    b
8108    n
8109    b
8110    b
8111    b
8112    b
8113    n
8114    b
8115    b
8116    n
8117    n
8118    n
8119    b
8120    b
8121    b
8122    n
8123    b
Name: Gill-size, Length: 8124, dtype: object

Узнаем общую информацию по нашей таблице.

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
Class                       8124 non-null object
Cap-shape                   8124 non-null object
Cap-surface                 8124 non-null object
Cap-color                   8124 non-null object
Bruises                     8124 non-null object
Odor                        8124 non-null object
Gill-attachment             8124 non-null object
Gill-spacing                8124 non-null object
Gill-size                   8124 non-null object
Gill-color                  8124 non-null object
Stalk-shape                 8124 non-null object
Stalk-root                  5644 non-null object
Stalk-surface-above-ring    8124 non-null object
Stalk-surface-below-ring    8124 non-null object
Stalk-color-above-ring      8124 non-null object
Stalk-color-below-ring      8124 non-null object
Veil-type                   8124 non-null object
Veil-color                  8124 non-null object
Ring-number

Вывод первых пяти элементов указанных нами столбцов.

In [12]:
data[['Class', 'Population']].head()

Class Population
0     p          s
1     e          n
2     e          n
3     p          s
4     e          a

Выведем основные статистические характеристики данных нашей таблицы.

In [13]:
data.describe()

Class Cap-shape Cap-surface Cap-color Bruises  Odor Gill-attachment  \
count   8124      8124        8124      8124    8124  8124            8124   
unique     2         6           4        10       2     9               2   
top        e         x           y         n       f     n               f   
freq    4208      3656        3244      2284    4748  3528            7914   

       Gill-spacing Gill-size Gill-color   ...   Stalk-surface-below-ring  \
count          8124      8124       8124   ...                       8124   
unique            2         2         12   ...                          4   
top               c         b          b   ...                          s   
freq           6812      5612       1728   ...                       4936   

       Stalk-color-above-ring Stalk-color-below-ring Veil-type Veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1          4   
top                         w                      w         p          w   
freq                     4464                   4384      8124       7924   

       Ring-number Ring-type Spore-print-color Population Habitat  
count         8124      8124              8124       8124    8124  
unique           3         5                 9          6       7  
top              o         p                 w          v       d  
freq          7488      3968              2388       4040    3148  

[4 rows x 23 columns]

где count – количество элементов в каждом столбце; unique – количество уникальных значений в столбце; top – наиболее часто встречающееся значение; freq – количество top значения в столбце.

Проверим элементы на соответствие заданной характеристике.

In [14]:
data['Class'] == 'e'

0       False
1        True
2        True
3       False
4        True
5        True
6        True
7        True
8       False
9        True
10       True
11       True
12       True
13      False
14       True
15       True
16       True
17      False
18      False
19      False
20       True
21      False
22       True
23       True
24       True
25      False
26       True
27       True
28       True
29       True
        ...  
8094     True
8095    False
8096     True
8097    False
8098    False
8099     True
8100     True
8101    False
8102     True
8103     True
8104     True
8105     True
8106     True
8107     True
8108    False
8109     True
8110     True
8111     True
8112     True
8113    False
8114    False
8115     True
8116    False
8117    False
8118    False
8119     True
8120     True
8121     True
8122    False
8123     True
Name: Class, Length: 8124, dtype: bool

Грибы со значением true являются съедобными, а со значением false – несъедобными.

## Готовим данные

Узнаем категорию данных одного столбца.

In [15]:
data['Ring-type'].dtype.name 

'object'

Разделим имеющиеся столбцы на категориальные и на числовые.

In [16]:
categorical_columns = [c for c in data.columns if data[c].dtype.name == 'object']
numerical_columns   = [c for c in data.columns if data[c].dtype.name != 'object']
print(categorical_columns)
print(numerical_columns)

['Class', 'Cap-shape', 'Cap-surface', 'Cap-color', 'Bruises', 'Odor', 'Gill-attachment', 'Gill-spacing', 'Gill-size', 'Gill-color', 'Stalk-shape', 'Stalk-root', 'Stalk-surface-above-ring', 'Stalk-surface-below-ring', 'Stalk-color-above-ring', 'Stalk-color-below-ring', 'Veil-type', 'Veil-color', 'Ring-number', 'Ring-type', 'Spore-print-color', 'Population', 'Habitat']
[]


В моей таблице все столбцы являются категориальными.

Для каждого столбца определим возможные значения.

In [17]:
for c in categorical_columns:
    print(c, data[c].unique())

Class ['p' 'e']
Cap-shape ['x' 'b' 's' 'f' 'k' 'c']
Cap-surface ['s' 'y' 'f' 'g']
Cap-color ['n' 'y' 'w' 'g' 'e' 'p' 'b' 'u' 'c' 'r']
Bruises ['t' 'f']
Odor ['p' 'a' 'l' 'n' 'f' 'c' 'y' 's' 'm']
Gill-attachment ['f' 'a']
Gill-spacing ['c' 'w']
Gill-size ['n' 'b']
Gill-color ['k' 'n' 'g' 'p' 'w' 'h' 'u' 'e' 'b' 'r' 'y' 'o']
Stalk-shape ['e' 't']
Stalk-root ['e' 'c' 'b' 'r' nan]
Stalk-surface-above-ring ['s' 'f' 'k' 'y']
Stalk-surface-below-ring ['s' 'f' 'y' 'k']
Stalk-color-above-ring ['w' 'g' 'p' 'n' 'b' 'e' 'o' 'c' 'y']
Stalk-color-below-ring ['w' 'p' 'g' 'b' 'n' 'e' 'y' 'o' 'c']
Veil-type ['p']
Veil-color ['w' 'n' 'o' 'y']
Ring-number ['o' 't' 'n']
Ring-type ['p' 'e' 'l' 'f' 'n']
Spore-print-color ['k' 'n' 'u' 'h' 'w' 'r' 'o' 'y' 'b']
Population ['s' 'n' 'a' 'v' 'y' 'c']
Habitat ['u' 'g' 'm' 'd' 'p' 'w' 'l']


## Обрабатываем пропущенные значения

In [18]:
data[categorical_columns].count(axis = 0)

Class                       8124
Cap-shape                   8124
Cap-surface                 8124
Cap-color                   8124
Bruises                     8124
Odor                        8124
Gill-attachment             8124
Gill-spacing                8124
Gill-size                   8124
Gill-color                  8124
Stalk-shape                 8124
Stalk-root                  5644
Stalk-surface-above-ring    8124
Stalk-surface-below-ring    8124
Stalk-color-above-ring      8124
Stalk-color-below-ring      8124
Veil-type                   8124
Veil-color                  8124
Ring-number                 8124
Ring-type                   8124
Spore-print-color           8124
Population                  8124
Habitat                     8124
dtype: int64

В столбце Stalk-root есть пропущенные значения. Заменим их на самое популярное значение в этом столбце.

In [19]:
data['Stalk-root'].describe().top

'b'

In [20]:
data['Stalk-root'] = data['Stalk-root'].fillna('b')

Ещё раз проверим пропущенные значения.

In [21]:
data[categorical_columns].count(axis = 0)

Class                       8124
Cap-shape                   8124
Cap-surface                 8124
Cap-color                   8124
Bruises                     8124
Odor                        8124
Gill-attachment             8124
Gill-spacing                8124
Gill-size                   8124
Gill-color                  8124
Stalk-shape                 8124
Stalk-root                  8124
Stalk-surface-above-ring    8124
Stalk-surface-below-ring    8124
Stalk-color-above-ring      8124
Stalk-color-below-ring      8124
Veil-type                   8124
Veil-color                  8124
Ring-number                 8124
Ring-type                   8124
Spore-print-color           8124
Population                  8124
Habitat                     8124
dtype: int64

Теперь основные статистические характеристики нашей таблицы выглядят так:

In [22]:
data.describe()

Class Cap-shape Cap-surface Cap-color Bruises  Odor Gill-attachment  \
count   8124      8124        8124      8124    8124  8124            8124   
unique     2         6           4        10       2     9               2   
top        e         x           y         n       f     n               f   
freq    4208      3656        3244      2284    4748  3528            7914   

       Gill-spacing Gill-size Gill-color   ...   Stalk-surface-below-ring  \
count          8124      8124       8124   ...                       8124   
unique            2         2         12   ...                          4   
top               c         b          b   ...                          s   
freq           6812      5612       1728   ...                       4936   

       Stalk-color-above-ring Stalk-color-below-ring Veil-type Veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1          4   
top                         w                      w         p          w   
freq                     4464                   4384      8124       7924   

       Ring-number Ring-type Spore-print-color Population Habitat  
count         8124      8124              8124       8124    8124  
unique           3         5                 9          6       7  
top              o         p                 w          v       d  
freq          7488      3968              2388       4040    3148  

[4 rows x 23 columns]

In [23]:
data_describe = data.describe(include = [object])
for c in categorical_columns:
    data[c] = data[c].fillna(data_describe[c]['top'])

## Векторизация

In [24]:
data_describe = data.describe(include = [object])
binary_columns = [c for c in categorical_columns if data_describe[c]['unique'] == 2]
nonbinary_columns = [c for c in categorical_columns if data_describe[c]['unique'] > 2]
print(binary_columns, nonbinary_columns)

['Class', 'Bruises', 'Gill-attachment', 'Gill-spacing', 'Gill-size', 'Stalk-shape'] ['Cap-shape', 'Cap-surface', 'Cap-color', 'Odor', 'Gill-color', 'Stalk-root', 'Stalk-surface-above-ring', 'Stalk-surface-below-ring', 'Stalk-color-above-ring', 'Stalk-color-below-ring', 'Veil-color', 'Ring-number', 'Ring-type', 'Spore-print-color', 'Population', 'Habitat']


Бинарные столбцы Class, Bruises, Gill-attachment, Gill-spacing, Gill-size, Stalk-shape.

Столбец Class я не трогаю, т.к. буду предсказывать съедобность гриба.
Посмотрим, какие уникальные значения принимает столбец Bruises.

In [27]:
data['Bruises'].unique()

array(['t', 'f'], dtype=object)

Заменим их на 0 и 1.

In [28]:
data.at[data['Bruises'] == 't', 'Bruises'] = 0
data.at[data['Bruises'] == 'f', 'Bruises'] = 1
data['Bruises'].describe()


count     8124
unique       2
top          1
freq      4748
Name: Bruises, dtype: int64

Аналогично для других бинарных столбцов.

In [29]:
data['Gill-attachment'].unique()

array(['f', 'a'], dtype=object)

In [30]:
data.at[data['Gill-attachment'] == 'f', 'Gill-attachment'] = 0
data.at[data['Gill-attachment'] == 'a', 'Gill-attachment'] = 1
data['Gill-attachment'].describe()

count     8124
unique       2
top          0
freq      7914
Name: Gill-attachment, dtype: int64

In [31]:
data['Gill-spacing'].unique()

array(['c', 'w'], dtype=object)

In [32]:
data.at[data['Gill-spacing'] == 'c', 'Gill-spacing'] = 0
data.at[data['Gill-spacing'] == 'w', 'Gill-spacing'] = 1
data['Gill-spacing'].describe()

count     8124
unique       2
top          0
freq      6812
Name: Gill-spacing, dtype: int64

In [33]:
data['Gill-size'].unique()

array(['n', 'b'], dtype=object)

In [34]:
data.at[data['Gill-size'] == 'n', 'Gill-size'] = 0
data.at[data['Gill-size'] == 'b', 'Gill-size'] = 1
data['Gill-size'].describe()

count     8124
unique       2
top          1
freq      5612
Name: Gill-size, dtype: int64

In [35]:
data['Stalk-shape'].unique()

array(['e', 't'], dtype=object)

In [36]:
data.at[data['Stalk-shape'] == 'e', 'Stalk-shape'] = 0
data.at[data['Stalk-shape'] == 't', 'Stalk-shape'] = 1
data['Stalk-shape'].describe()

count     8124
unique       2
top          1
freq      4608
Name: Stalk-shape, dtype: int64

Воспользуемся методом get_dummies для векторизации оставшихся признаков.

In [37]:
data_nonbinary = pd.get_dummies(data[nonbinary_columns])
print(data_nonbinary.columns)

Index(['Cap-shape_b', 'Cap-shape_c', 'Cap-shape_f', 'Cap-shape_k',
       'Cap-shape_s', 'Cap-shape_x', 'Cap-surface_f', 'Cap-surface_g',
       'Cap-surface_s', 'Cap-surface_y',
       ...
       'Population_s', 'Population_v', 'Population_y', 'Habitat_d',
       'Habitat_g', 'Habitat_l', 'Habitat_m', 'Habitat_p', 'Habitat_u',
       'Habitat_w'],
      dtype='object', length=105)


### Соединяем в одну таблицу бинарные и не бинарные признаки.

In [38]:
data = pd.concat((data_nonbinary, data[binary_columns]), axis = 1)
print(data.shape)

(8124, 111)


In [39]:
data.describe()

Cap-shape_b  Cap-shape_c  Cap-shape_f  Cap-shape_k  Cap-shape_s  \
count  8124.000000  8124.000000  8124.000000  8124.000000  8124.000000   
mean      0.055638     0.000492     0.387986     0.101920     0.003939   
std       0.229235     0.022185     0.487321     0.302562     0.062641   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     1.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

       Cap-shape_x  Cap-surface_f  Cap-surface_g  Cap-surface_s  \
count  8124.000000    8124.000000    8124.000000    8124.000000   
mean      0.450025       0.285574       0.000492       0.314623   
std       0.497527       0.451715       0.022185       0.464394   
min       0.000000       0.000000       0.000000       0.000000   
25%       0.000000       0.000000       0.000000       0.000000   
50%       0.000000       0.000000       0.000000       0.000000   
75%       1.000000       1.000000       0.000000       1.000000   
max       1.000000       1.000000       1.000000       1.000000   

       Cap-surface_y     ...       Population_s  Population_v  Population_y  \
count    8124.000000     ...        8124.000000   8124.000000   8124.000000   
mean        0.399311     ...           0.153619      0.497292      0.210734   
std         0.489787     ...           0.360605      0.500023      0.407855   
min         0.000000     ...           0.000000      0.000000      0.000000   
25%         0.000000     ...           0.000000      0.000000      0.000000   
50%         0.000000     ...           0.000000      0.000000      0.000000   
75%         1.000000     ...           0.000000      1.000000      0.000000   
max         1.000000     ...           1.000000      1.000000      1.000000   

         Habitat_d    Habitat_g    Habitat_l    Habitat_m    Habitat_p  \
count  8124.000000  8124.000000  8124.000000  8124.000000  8124.000000   
mean      0.387494     0.264402     0.102413     0.035943     0.140817   
std       0.487208     0.441041     0.303209     0.186159     0.347854   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     1.000000     0.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

         Habitat_u    Habitat_w  
count  8124.000000  8124.000000  
mean      0.045298     0.023634  
std       0.207969     0.151914  
min       0.000000     0.000000  
25%       0.000000     0.000000  
50%       0.000000     0.000000  
75%       0.000000     0.000000  
max       1.000000     1.000000  

[8 rows x 105 columns]

В моей таблице есть характеристика "Class", которая показывает съедобный гриб или нет.

Я буду прогнозировать съедобность по остальным признакам, описывающим внешний вид гриба.

In [40]:
X = data.drop(('Class'), axis = 1)
y = data['Class']
feature_names = X.columns

In [41]:
type(feature_names)

pandas.core.indexes.base.Index

In [42]:
feature_names

Index(['Cap-shape_b', 'Cap-shape_c', 'Cap-shape_f', 'Cap-shape_k',
       'Cap-shape_s', 'Cap-shape_x', 'Cap-surface_f', 'Cap-surface_g',
       'Cap-surface_s', 'Cap-surface_y',
       ...
       'Habitat_l', 'Habitat_m', 'Habitat_p', 'Habitat_u', 'Habitat_w',
       'Bruises', 'Gill-attachment', 'Gill-spacing', 'Gill-size',
       'Stalk-shape'],
      dtype='object', length=110)

In [43]:
print(X.shape)
print(y.shape)
N, d = X.shape

(8124, 110)
(8124,)


In [44]:
type(X)

pandas.core.frame.DataFrame

In [45]:
type(y)

pandas.core.series.Series

In [46]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)

N_train, _ = X_train.shape 
N_test,  _ = X_test.shape 

print(N_train, N_test)

5686 2438


In [47]:
5686 + 2438

8124

## kNN

In [48]:
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors = 7)
knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=7, p=2,
           weights='uniform')

In [49]:
knn

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=7, p=2,
           weights='uniform')

In [51]:
err_test = 1 - knn.score(X_test, y_test)

In [52]:
err_test

0.0

In [53]:
from sklearn.metrics import confusion_matrix

print(confusion_matrix(y_test, y_test_predict))

[[1257    0]
 [   0 1181]]


In [54]:
confusion_matrix([0,1,0,1,0], [0,1,0,0,0])

array([[3, 0],
       [1, 1]], dtype=int64)

## Подбор параметров

In [55]:
from sklearn.model_selection import GridSearchCV
nnb = [1, 3, 5, 10, 15, 20, 25, 35, 45, 55]
knn = KNeighborsClassifier()
grid = GridSearchCV(knn, param_grid = {'n_neighbors': nnb}, cv=10)
grid.fit(X_train, y_train)

best_cv_err = 1 - grid.best_score_
best_n_neighbors = grid.best_estimator_.n_neighbors
print(best_cv_err, best_n_neighbors)

0.0 1


Метод подобрал k = 1

In [56]:
knn = KNeighborsClassifier(n_neighbors = best_n_neighbors).fit(X_train, y_train)

err_train = np.mean(y_train != knn.predict(X_train))
err_test  = np.mean(y_test  != knn.predict(X_test))

print(err_train, err_test)

0.0 0.0
